src/addons/keyboard_host.cpp

```diff
@@ -0,0 +1,197 @@
+  gamepad->state.buttons  |= _keyboard_host_state.buttons;
+  gamepad->state.lx       |= _keyboard_host_state.lx;
+  gamepad->state.ly       |= _keyboard_host_state.ly;
+  gamepad->state.rx       |= _keyboard_host_state.rx;
+  gamepad->state.ry       |= _keyboard_host_state.ry;
+  gamepad->state.lt       |= _keyboard_host_state.lt;
+  gamepad->state.rt       |= _keyboard_host_state.rt;
+
+  tuh_task();
+}
+
+// Invoked when device with hid interface is mounted
+// Report descriptor is also available for use. tuh_hid_parse_report_descriptor()
+// can be used to parse common/simple enough descriptor.
+// Note: if report descriptor length > CFG_TUH_ENUMERATION_BUFSIZE, it will be skipped
+// therefore report_desc = NULL, desc_len = 0
+void tuh_hid_mount_cb(uint8_t dev_addr, uint8_t instance, uint8_t const* desc_report, uint16_t desc_len)
+{
+  (void)desc_report;
+  (void)desc_len;
+
+  // Interface protocol (hid_interface_protocol_enum_t)
+  uint8_t const itf_protocol = tuh_hid_interface_protocol(dev_addr, instance);
+
+  uint16_t vid, pid;
+  tuh_vid_pid_get(dev_addr, &vid, &pid);
+
+  // tuh_hid_report_received_cb() will be invoked when report is available
+  if (itf_protocol == HID_ITF_PROTOCOL_KEYBOARD)
+  {
+    if ( !tuh_hid_receive_report(dev_addr, instance) )
+    {
+      // Error: cannot request report
+    }
+  }
+}
+
+// Invoked when device with hid interface is un-mounted
+void tuh_hid_umount_cb(uint8_t dev_addr, uint8_t instance)
+{
+  (void) dev_addr;
+  (void) instance;
+}
+
+uint8_t getKeycodeFromModifier(uint8_t modifier) {
+	switch (modifier) {
+	  case KEYBOARD_MODIFIER_LEFTCTRL   : return HID_KEY_CONTROL_LEFT ;
+	  case KEYBOARD_MODIFIER_LEFTSHIFT  : return HID_KEY_SHIFT_LEFT   ;
+	  case KEYBOARD_MODIFIER_LEFTALT    : return HID_KEY_ALT_LEFT     ;
+	  case KEYBOARD_MODIFIER_LEFTGUI    : return HID_KEY_GUI_LEFT     ;
+	  case KEYBOARD_MODIFIER_RIGHTCTRL  : return HID_KEY_CONTROL_RIGHT;
+	  case KEYBOARD_MODIFIER_RIGHTSHIFT : return HID_KEY_SHIFT_RIGHT  ;
+	  case KEYBOARD_MODIFIER_RIGHTALT   : return HID_KEY_ALT_RIGHT    ;
+	  case KEYBOARD_MODIFIER_RIGHTGUI   : return HID_KEY_GUI_RIGHT    ;
+	}
+
+	return 0;
+}
+
+// convert hid keycode to ascii and print via usb device CDC (ignore non-printable)
+void process_kbd_report(uint8_t dev_addr, hid_keyboard_report_t const *report)
+{
+  (void) dev_addr;
+
+  _keyboard_host_state.dpad = 0;
+  _keyboard_host_state.buttons = 0;
+  _keyboard_host_state.lx = GAMEPAD_JOYSTICK_MID;
+  _keyboard_host_state.ly = GAMEPAD_JOYSTICK_MID;
+  _keyboard_host_state.rx = GAMEPAD_JOYSTICK_MID;
+  _keyboard_host_state.ry = GAMEPAD_JOYSTICK_MID; 
+  _keyboard_host_state.lt = 0;
+  _keyboard_host_state.rt = 0;
+
+  for(uint8_t i=0; i<7; i++)
+  {
+    uint8_t keycode = (i < 6) ? report->keycode[i] : getKeycodeFromModifier(report->modifier);
+    if ( keycode )
+    {
+      GamepadOptions& gamepadOptions = Storage::getInstance().getGamepadOptions();
```

Comment:
This can be a `const` reference. You are not modifying the `struct` behind the reference.